src/NHibernate/Engine/ForeignKeys.cs

```diff
@@ -171,6 +171,9 @@ public static bool IsNotTransientSlow(string entityName, object entity, ISession
 		/// </remarks>
 		public static bool? IsTransientFast(string entityName, object entity, ISessionImplementor session)
 		{
+			if (entity.IsProxy())
+				return false;
```

Comment:
So, despite some [hesitations](https://github.com/nhibernate/nhibernate-core/pull/2045#discussion_r263922575) on that subject, now proxies are treated as never transient.

What happens if we load a proxy of an entity with an assigned identifier, initialize it, delete it in session, flush, and save it again? In such case we would have a transient proxy, and it should be possible to save it, I guess. If that case does work with currently released NHibernate, then the logic from "not transient" should not be applied to "transient" too. (And we should consider removing it from the "not" case.)

Reply:
This scenario still works (both direct and cascace saving) - as entity is unproxied along the way to `IsTransientFast`. I will commit tests for such cases later.

But from correctness standpoint you are right - my initial implementation is more correct. It correctly detects transient objects from any context. So I think I simply need to drop all my commits to the state you've approved initially. Do you agree? That means that `GetIdentifier` fix is also required here.

Reply:
Your initial implementation has my preference.

@hazzik,

> I would prefer to keep this check outside of this method as in other places.

Is-it because this is a patch of  5.2.x? So you consider that this kind of (even small) "semantic" change should be avoided?
(I do not actually see it as a semantic change. I see it more like fixing a bug in `GetIdentifier`, since its documentation does not state it does not support proxies.)

Reply:
Sorry, I did not have time to clarify my position on this subject. While this check is ok here, I think it should be performed outside of this method. First of all, `AbstractEntityPersister`/`AbstractEntityTuplizer` does not deal with any kind of instantiaded proxies currently (except a delegated method to instantiate one). Secondly, `ISessionImplementor.GetEntityPersister(...)` is a very heavy method and really should be left as a last resort. So, check for the proxy should happen outside of the persister to avoid its lookup.

Reply:
>ISessionImplementor.GetEntityPersister(...) is a very heavy method and really should be left as a last resort. 

Why single dictionary lookup is considered heavy? And in our case it's `persister.IsTransient` method that calls `GetIdentifier`. So we really save nothing.
It's either put this check  in `persister.IsTransient` vs `GetIdentifier` (persister or tupilizer). To me the deeper the better - to cover all possible future cases (so in tupilizer).

> `AbstractEntityPersister/AbstractEntityTuplizer` does not deal with any kind of instantiaded proxies currently

I take it you meant - non-instantiated. As I already tell EntityPersister **must** be able to work with such proxy. For now it's only for `IsTransient` method.

Reply:
> Why single dictionary lookup is considered heavy?

You are confusing it with another method that is of `ISessionFactory`

Reply:
Ok. It could be heavy for subclasses. But that wasn't really the point. My main point for proper `IsTransient` detection we need to support proxies. And it's better to do right in `GetIdentifier` method

Reply:
I've pushed changes to the point I think is the right fix. Feel free to adjust it your way